### PR TITLE
Casecmp order

### PR DESCRIPTION
**Pull request by rrosenblum**
This should be merged after, or instead of, #2692.

### CODE REVIEW

**Comment by jonas054** [comment]
This fixes #2664 so add that in the commit message. But as I wrote in that issue, there are a few more problems that need fixing.

If I change the lower case `"a"` to an upper case `"A"` in one of the new spec examples:

```ruby
    it "autocorrects str.#{selector} == a variable" do
      new_source = autocorrect_source(cop, ['other = "A"',
                                            'str.downcase == other'])
      expect(new_source).to eq(['other = "A"',
                                'str.casecmp(other).zero?'].join("\n"))
    end
```
it still passes, but that's an auto-correction that changes the semantics of the code.

What I think we should do is to only report an offense if we know that the other value (the one that's not a result of `#downcase` or `#upcase`) has the same case. For example:
* `str.downcase == other.downcase`
* `str.upcase == 'A'`

Then we need a new lint cop that warns about expressions such as `str.upcase == 'a'` always being false.

**Comment by alexdowad** [comment]
> Then we need a new lint cop that warns about expressions such as str.upcase == 'a' always being false.

This is just one case of thousands where knowledge of what a platform method does can tell you that some expression is redundant. I don't think we can catalog every such case, though. And this particular case does not seem specially significant.

**Comment by jonas054** [comment]
> And this particular case does not seem specially significant.

Yes. perhaps you're right about that. A new lint cop might never get anything to report, as it should be a rare phenomenon.

**Comment by rrosenblum** [comment]
I agree that there are some oddities about the semantics of how this cop works, but it seems as though all oddities are a result of the initial code being bad. 

From the earlier example:
```ruby
    it "autocorrects str.#{selector} == a variable" do
      new_source = autocorrect_source(cop, ['other = "A"',
                                            'str.downcase == other'])
      expect(new_source).to eq(['other = "A"',
                                'str.casecmp(other).zero?'].join("\n"))
    end
```

This correction will always return false, but that is because the initial code will always return false. It would be nice to be able to warn users about ineffective code, but we would need to come up with a system that is able to infer what kind of object is stored to each variable. Right now, we have no idea what `other` is, but if we could scan the context of the code, we could determine that it is a string. This could lead to a number of cops being more useful. I have been trying to think of a way to implement a variable lookup for a little while now. My biggest concerns are maintaining context and performance.

**Comment by rrosenblum** [comment]
> This fixes #2664 so add that in the commit message. But as I wrote in that issue, there are a few more problems that need fixing.

Thanks, I didn't know there was an issue for it already. #2692 fixes the issue. This PR adds support for `'a' == str.downcase` and handles `str.downcase == obj.downcase`. I have updated the CHANGELOG accordingly.

**Comment by jonas054** [comment]
> it seems as though all oddities are a result of the initial code being bad.

Yes, this has nothing to do with your changes. Sorry I didn't mention that.

File: lib/rubocop/cop/performance/casecmp.rb [hunk 1]
```diff
@@ -9,49 +9,82 @@ module Performance
...
         END
 
         def on_send(node)
-          downcase_eq(node) do |send_downcase, case_method, eq_method|
-            range = node.loc.selector.join(send_downcase.loc.selector)
+          return if part_of_ignored_node?(node)
+
+          downcase_eq(node) do |send_downcase, case_method, eq_method, other|
+            *_, method = *other
+            if CASE_METHODS.include?(method)
```
Comment (DNNX) [inline lib/rubocop/cop/performance/casecmp.rb]: It is possible to inline `other` using _clojuresque_ destructuring:

```ruby
downcase_eq(node) do |send_downcase, case_method, eq_method, (*, method)|
  if CASE_METHODS.include?(method)
  ...
```

To me it's slightly more readable, but maybe it's just me. I don't have strong opinion about that.

What else bothers me a bit is the name of the variable `method`. It is not necessary a method, is it? Maybe it should be called `maybe_method`?

Or alternatively, we can leave `other` as is and extract `case_method_call?(expr)` method:

```ruby
downcase_eq(node) do |send_downcase, case_method, eq_method, other|
  if case_method_call?(other)
    range = ...
 ...

def case_method_call?(expr) # should it be def_node_matcher?
  CASE_METHODS.include?(expr.last)
end
```

**Comment by rrosenblum** [comment]
> It is possible to inline other using clojuresque destructuring:
```ruby
downcase_eq(node) do |send_downcase, case_method, eq_method, (*, method)|
```

I like this idea. Unfortunately, this did not produce the same effect as handling the expansion on a separate line, and the line length became too long.

> What else bothers me a bit is the name of the variable `method`. It is not necessary a method, is it? Maybe it should be called `maybe_method`?

This occurs in many places throughout the RuboCop source. I feel like the majority of the code uses naming that is going to assume the type of the object that we want in the particular place. If we expected `method` to be a variable, it would probably be named `variable` instead of `maybe_variable`

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@
 * [#2657](https://github.com/bbatsov/rubocop/pull/2657): Floating headers in HTML output. ([@mattparlane][])
 * Add new `Style/SpaceAroundKeyword` cop. ([@lumeet][])
 * [#2745](https://github.com/bbatsov/rubocop/pull/2745): New cop `Style/MultilineHashBraceLayout` checks that the closing brace in a hash literal is symmetrical with respect to the opening brace and the hash elements. ([@panthomakos][])
+* [#2699](https://github.com/bbatsov/rubocop/pull/2699): `Performance/Casecmp` can register offenses offenses when `str.downcase` or `str.upcase` are passed to an equality method. ([@rrosenblum][])
```
Comment (mvidner) [inline CHANGELOG.md]: "offenses offenses"

**Comment by bbatsov** [comment]
Has to be rebased.

**Pull request merged**